parlai/agents/rag/retrievers.py

```diff
@@ -0,0 +1,1217 @@
+            # when initializing from parlai rag models, the pretrained path
+            # may not longer exist. This is fine if we've alread trained
+            # the model.
+            assert dpr_model == 'bert_from_parlai_rag'
+            logging.error(f'Pretrained Path does not exist: {pretrained_path}')
+            pretrained_path = modelzoo_path(opt['datapath'], DPR_ZOO_MODEL)
+            dpr_model = 'bert'
+            logging.error(f'Setting to zoo model: {pretrained_path}')
+        enc_opt = {
+            "n_heads": config.num_attention_heads,
+            "n_layers": config.num_hidden_layers,
+            "embedding_size": config.hidden_size,
+            "ffn_size": config.intermediate_size,
+            "dropout": config.hidden_dropout_prob,
+            "attention_dropout": config.attention_probs_dropout_prob,
+            "activation": config.hidden_act,
+            "variant": 'xlm',
+            "reduction_type": 'first',
+            "n_positions": config.max_position_embeddings,
+            "n_segments": config.type_vocab_size,
+        }
+        embedding = torch.nn.Embedding(
+            config.vocab_size, config.hidden_size, padding_idx=config.pad_token_id
+        )
+        super().__init__(
+            enc_opt,
+            vocabulary_size=config.vocab_size,
+            padding_idx=config.pad_token_id,
+            embedding=embedding,
+            reduction_type='first',
+        )
+
+        self._load_state(dpr_model, pretrained_path, encoder_type)
+
+    def _load_state(self, dpr_model: str, pretrained_path: str, encoder_type: str):
+        """
+        Load pre-trained model states.
+
+        :param dpr_model:
+            which dpr model type we're using
+        :param pretrained_path:
+            path to pretrained model
+        :param encoder_type:
+            whether this is a query or document encoder
+        """
+        if dpr_model == 'bert':
+            state_dict = BertConversionUtils.load_bert_state(
+                self.state_dict(),
+                pretrained_dpr_path=pretrained_path,
+                encoder_type=encoder_type,
+            )
+            self.load_state_dict(state_dict)
+        elif dpr_model == 'bert_from_parlai_rag':
+            state_dict = torch.load(pretrained_path, map_location='cpu')["model"]
+            key = f"{encoder_type}_encoder."
+            state_dict = {
+                k.split(key)[-1]: v for k, v in state_dict.items() if key in k
+            }
+            self.load_state_dict(state_dict)
+
+
+class RagDprQueryEncoder(RagDprEncoder):
+    """
+    Query Encoder for DPR.
+    """
+
+    def __init__(self, *args, **kwargs):
+        kwargs['encoder_type'] = 'query'
+        super().__init__(*args, **kwargs)
+
+
+class RagDprDocumentEncoder(RagDprEncoder):
+    """
+    Document Encoder for DPR.
+    """
+
+    def __init__(self, *args, **kwargs):
+        kwargs['encoder_type'] = 'document'
+        super().__init__(*args, **kwargs)
```

Comment:
It kind of feels like all of this could/should live in `RagModel.build_encoder`. Something like:
```
    @classmethod
    def build_encoder(
        cls,
        opt: Opt,
        *args,
        dictionary: Optional[DictionaryAgent] = None,
        embedding: Optional[torch.nn.Embedding] = None,
        encoder_class: Optional[Type] = None,
        **kwargs,
    ):
        # Override options
        config: BertConfig = BertConfig.from_pretrained('bert-base-uncased')
        pretrained_path = modelzoo_path(opt['datapath'], pretrained_path)
        if not os.path.exists(pretrained_path):
            # when initializing from parlai rag models, the pretrained path
            # may not longer exist. This is fine if we've alread trained
            # the model.
            assert dpr_model == 'bert_from_parlai_rag'
            logging.error(f'Pretrained Path does not exist: {pretrained_path}')
            pretrained_path = modelzoo_path(opt['datapath'], DPR_ZOO_MODEL)
            dpr_model = 'bert'
            logging.error(f'Setting to zoo model: {pretrained_path}')
        enc_opt = {
            "n_heads": config.num_attention_heads,
            "n_layers": config.num_hidden_layers,
            "embedding_size": config.hidden_size,
            "ffn_size": config.intermediate_size,
            "dropout": config.hidden_dropout_prob,
            "attention_dropout": config.attention_probs_dropout_prob,
            "activation": config.hidden_act,
            "variant": 'xlm',
            "reduction_type": 'first',
            "n_positions": config.max_position_embeddings,
            "n_segments": config.type_vocab_size,
        }
        encoder = TransformerEncoder(
            enc_opt,
            vocabulary_size=config.vocab_size,
            padding_idx=config.pad_token_id,
            reduction_type='first',
            **kwargs,
        )
        if dpr_model == 'bert':
            state_dict = BertConversionUtils.load_bert_state(
                self.state_dict(),
                pretrained_dpr_path=pretrained_path,
                encoder_type=encoder_type,
            )
            encoder.load_state_dict(state_dict)
        elif dpr_model == 'bert_from_parlai_rag':
            state_dict = torch.load(pretrained_path, map_location='cpu')["model"]
            key = f"{encoder_type}_encoder."
            state_dict = {
                k.split(key)[-1]: v for k, v in state_dict.items() if key in k
            }
            encoder.load_state_dict(state_dict)
        return encoder
```

Reply:
im not sure what you mean here --> the `RagDprEncoder` is used exclusively in the retriever modules, we don't touch it in `RagModel`

Reply:
Ah, regarding `RagModel` I got a little confused because the words "encoder" and "decoder" are used for multiple things in that model.

What I was getting at, though, is that `RagDprDocumentEncoder`, `RagDprQueryEncoder`, and `RagDprEncoder` don't do much. And what they _do_ do could just as well happen outside of the class instead of inside.

So instead of creating new classes that don't do much (since classes/objects come along with state implications and some overhead) you could just do everything that's happening in these classes in `build_encoder`. Well with your latest changes you'd do it in `build_context_encoder` and `build_candidate_encoder`.

Anyway, that might save a little code but doesn't change muc functionally. So if it doesn't feel intuitive or cleaner to you then don't worry about it!

Reply:
ahh, yes, i see your point. while these are built in `build_context_encoder` and `build_candidate_encoder` for the `DprModel` (when considering DPR as its own agent), they are instantiated directly when used within the `DprRetriever`, which doesn't really have a concept of `build_encoder`. so, i think i'll keep it as is, but I definitely understand where you're coming from